### PR TITLE
:computer: :dart: Pressure smoothing with MPI

### PR DESCRIPTION
**Pull request by kks32**
Add Ezra's implementation of pressure smoothing fixed MPI #324

### CODE REVIEW

File: include/cell.tcc [hunk 1]
```diff
@@ -1009,6 +1009,17 @@ void mpm::Cell<Tdim>::map_mass_momentum_to_nodes(
...
 
+//! Map particle pressure to nodes for a given phase
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::map_pressure_to_nodes(const Eigen::VectorXd& shapefn,
+                                            unsigned phase, double pmass,
+                                            double ppressure) {
+
+  for (unsigned i = 0; i < this->nfunctions(); ++i)
+    nodes_[i]->update_pressure(true, phase,
+                               shapefn(i) * pmass * shapefn(i) * ppressure);
```
Comment (srhgk2) [inline include/cell.tcc]: Is `shapefn(i) * pmass * shapefn(i) * ppressure` correct? Why do we have two `shapefn(i)`?

File: include/particle.tcc [hunk 2]
```diff
@@ -593,3 +593,48 @@ bool mpm::Particle<Tdim, Tnphases>::update_pressure(unsigned phase,
...
+bool mpm::Particle<Tdim, Tnphases>::map_pressure_to_nodes(unsigned phase) {
+  bool status = true;
+  try {
+    // Check if particle mass is set
+    if (mass_(phase) != std::numeric_limits<double>::max()) {
+      // Map particle mass and momentum to nodes
+      this->cell_->map_pressure_to_nodes(this->shapefn_, phase, mass_(phase),
+                                         pressure_(phase));
+    } else {
+      throw std::runtime_error("Particle mass has not be computed");
```
Comment (srhgk2) [inline include/particle.tcc]: may be "has not been computed"?

File: include/particle.tcc [hunk 3]
```diff
@@ -593,3 +593,48 @@ bool mpm::Particle<Tdim, Tnphases>::update_pressure(unsigned phase,
...
+
+// Compute pressure smoothing of the particle based on nodal pressure
+template <unsigned Tdim, unsigned Tnphases>
+bool mpm::Particle<Tdim, Tnphases>::compute_pressure_smoothing(unsigned phase) {
+  bool status = true;
+  try {
+    // Check if particle has a valid cell ptr
+    if (cell_ != nullptr) {
+      // Get interpolated nodal pressure
+      double pressure =
```
Comment (srhgk2) [inline include/particle.tcc]: may be we don't need this copy.

File: tests/cell_test.cc [hunk 4]
```diff
@@ -617,6 +619,11 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
         for (unsigned i = 0; i < pvelocity.size(); ++i)
           REQUIRE(node->momentum(phase)(i) == Approx(2.0).epsilon(Tolerance));
       }
+
+      // Map particle pressure to nodes
+      cell->map_pressure_to_nodes(shapefns_xi, phase, pmass, ppressure);
+      for (const auto& node : nodes)
+        REQUIRE(node->pressure(phase) == Approx(1.5).epsilon(Tolerance));
```
Comment (srhgk2) [inline tests/cell_test.cc]: `Approx(1.5)` may change depending on how we map pressure to nodes

File: tests/cell_test.cc [hunk 5]
```diff
@@ -1835,6 +1897,11 @@ TEST_CASE("Cell is checked for 3D case", "[cell][3D]") {
         for (unsigned i = 0; i < pvelocity.size(); ++i)
           REQUIRE(node->momentum(phase)(i) == Approx(1.0).epsilon(Tolerance));
       }
+
+      // Map particle pressure to nodes
+      cell->map_pressure_to_nodes(shapefns_xi, phase, pmass, ppressure);
+      for (const auto& node : nodes)
+        REQUIRE(node->pressure(phase) == Approx(0.75).epsilon(Tolerance));
```
Comment (srhgk2) [inline tests/cell_test.cc]: same as 2d

File: include/cell.tcc (hunk 1 above)
Comment (kks32) [reply include/cell.tcc]: I was modifying Ezra's equations to do MPI. So I originally had `shapefn(i) * pmass * ppressure`, but that failed the test case. I think what this is trying to do is map particle mass and pressure separately to the nodes, then update the pressure. I need to check the original equation

Comment (kks32) [reply include/cell.tcc]: Having looked at the equation, I think it should be `shapefn(i) * pmass * ppressure`

**Review by srhgk2 — APPROVED** [review]

**Pull request merged**